### PR TITLE
PWGHF: Add more checks on collision association probability

### PR DESCRIPTION
**Pull request by fgrosa**

### CODE REVIEW

File: PWGHF/Tasks/HFMCValidation.cxx [hunk 1]
```diff
@@ -278,15 +311,27 @@ struct ValidationRecLevel {
...
   {
     // loop over tracks
     for (auto& track : tracks) {
       uint index = uint(track.collisionId() >= 0);
       if (track.has_mcParticle()) {
         auto particle = track.mcParticle(); // get corresponding MC particle to check origin
+        uint index2 = 0;
+        if (index) {
+          auto collision = collisions.rawIteratorAt(track.collisionId());
+          auto mcCollision = mcCollisions.rawIteratorAt(particle.mcCollisionId());
```
Comment (vkucera) [inline PWGHF/Tasks/HFMCValidation.cxx:324]: Can you not get the collisions directly via the getters `track.collision()`, `particle.mcCollision()` (or `_as<>()`)?

File: PWGHF/Tasks/HFMCValidation.cxx [hunk 2]
```diff
@@ -278,15 +311,27 @@ struct ValidationRecLevel {
...
+          if (std::abs(collision.mcCollisionId() - particle.mcCollisionId()) == 0 && std::abs(collision.posZ() - mcCollision.posZ()) < 0.02) { // 200 microns compatibility of Z vertex position also required
```
Comment (vkucera) [inline PWGHF/Tasks/HFMCValidation.cxx:325]: I suppose you just want to compare the indices here.
```suggestion
          if (collision.mcCollisionId() == particle.mcCollisionId() && std::abs(collision.posZ() - mcCollision.posZ()) < 0.02) { // 200 microns compatibility of Z vertex position also required
```

File: PWGHF/Tasks/HFMCValidation.cxx (hunk 1 above)
Comment (fgrosa) [reply PWGHF/Tasks/HFMCValidation.cxx]: I did it for the `mcCollision` but if I do it also for the collision for some reason then I cannot call `mcCollisionId()` even if I joined the collision table with the one with MC collision labels (so I can only address your comment partially :-))

**Review by cterrevo — APPROVED** [review]

File: PWGHF/Tasks/HFMCValidation.cxx (hunk 1 above)
Comment (vkucera) [reply PWGHF/Tasks/HFMCValidation.cxx]: Are you saying that if you do `auto collision = track.collision_as<CollisionsWithMCLabels>();`, calling `collision.mcCollisionId()` then crashes?

Comment (fgrosa) [reply PWGHF/Tasks/HFMCValidation.cxx]: With `collision_as` indeed works

File: PWGHF/Tasks/HFMCValidation.cxx [hunk 3]
```diff
@@ -287,6 +321,15 @@ struct ValidationRecLevel {
         auto particle = track.mcParticle(); // get corresponding MC particle to check origin
         auto origin = RecoDecay::getCharmHadronOrigin(particlesMC, particle, true);
         histOriginTracks[index]->Fill(origin, track.pt());
+        if (index) {
+          auto collision = collisions.rawIteratorAt(track.collisionId());
+          auto mcCollision = particle.mcCollision();
+          uint index2 = 2;
+          if (collision.mcCollisionId() - particle.mcCollisionId() == 0 && std::abs(collision.posZ() - mcCollision.posZ()) < 0.02) { // 200 microns compatibility of Z vertex position also required
```
[the remark this answers is not in the document]
Comment (fgrosa) [reply PWGHF/Tasks/HFMCValidation.cxx:328]: The two are equivalent ;)
I have implemented these two style modifications in this PR https://github.com/AliceO2Group/O2Physics/pull/1035

[the remark this answers is not in the document]
Comment (vkucera) [reply PWGHF/Tasks/HFMCValidation.cxx:328]: Yes, it is just more readable with `Id == Id`. Thanks

**Pull request merged**